warehouse/integrations/deltalake-native/deltalake_native.go

```diff
@@ -0,0 +1,1356 @@
+	"DOUBLE":    "float",
+	"BOOLEAN":   "boolean",
+	"STRING":    "string",
+	"DATE":      "date",
+	"TIMESTAMP": "datetime",
+	"tinyint":   "int",
+	"smallint":  "int",
+	"int":       "int",
+	"bigint":    "int",
+	"decimal":   "float",
+	"float":     "float",
+	"double":    "float",
+	"boolean":   "boolean",
+	"string":    "string",
+	"date":      "date",
+	"timestamp": "datetime",
+}
+
+// excludeColumnsMap Columns you need to exclude
+// Since event_date is an auto generated column in order to support partitioning.
+// We need to ignore it during query generation.
+var excludeColumnsMap = map[string]bool{
+	"event_date": true,
+}
+
+var primaryKeyMap = map[string]string{
+	warehouseutils.UsersTable:      "id",
+	warehouseutils.IdentifiesTable: "id",
+	warehouseutils.DiscardsTable:   "row_id",
+}
+
+var errorsMappings = []model.JobError{
+	{
+		Type:   model.PermissionError,
+		Format: regexp.MustCompile(`UnauthorizedAccessException: PERMISSION_DENIED: User does not have READ FILES on External Location`),
+	},
+	{
+		Type:   model.PermissionError,
+		Format: regexp.MustCompile(`SecurityException: User does not have permission CREATE on CATALOG`),
+	},
+}
+
+type Credentials struct {
+	Host    string
+	Port    string
+	Path    string
+	Token   string
+	Catalog string
+	timeout time.Duration
+}
+
+type Deltalake struct {
+	DB                     *sql.DB
+	Namespace              string
+	ObjectStorage          string
+	Warehouse              model.Warehouse
+	Uploader               warehouseutils.Uploader
+	ConnectTimeout         time.Duration
+	Logger                 logger.Logger
+	Stats                  stats.Stats
+	LoadTableStrategy      string
+	EnablePartitionPruning bool
+}
+
+func Init() {
+	pkgLogger = logger.NewLogger().Child("warehouse").Child("deltalake-native")
+}
+
+func NewDeltalake() *Deltalake {
+	return &Deltalake{
+		Logger: pkgLogger,
+		Stats:  stats.Default,
+	}
+}
+
+func WithConfig(h *Deltalake, config *config.Config) {
+	h.LoadTableStrategy = config.GetString("Warehouse.deltalake.loadTableStrategy", mergeMode)
+	h.EnablePartitionPruning = config.GetBool("Warehouse.deltalake.enablePartitionPruning", true)
+}
+
+func columnsWithDataTypes(columns model.TableSchema, prefix string) string {
+	keys := warehouseutils.SortColumnKeysFromColumnMap(columns)
+	format := func(idx int, name string) string {
+		if _, ok := excludeColumnsMap[name]; ok {
+			return ""
+		}
+		if name == "received_at" {
+			generatedColumnSQL := "DATE GENERATED ALWAYS AS ( CAST(received_at AS DATE) )"
+			return fmt.Sprintf(`%s%s %s, %s%s %s`, prefix, name, dataTypesMap[columns[name]], prefix, "event_date", generatedColumnSQL)
+		}
+
+		return fmt.Sprintf(`%s%s %s`, prefix, name, dataTypesMap[columns[name]])
+	}
+	return warehouseutils.JoinWithFormatting(keys, format, ",")
+}
+
+func columnNames(columns []string) string {
+	return strings.Join(columns, ",")
+}
+
+func stagingColumnNames(keys []string) string {
+	format := func(idx int, str string) string {
```

Comment:
[question] why do we need the idx variable here?

Reply:
It has to do with the generic `warehouseutils.JoinWithFormatting`, which requires a function with (index, str)

We could do:
```suggestion
	format := func(_ int, str string) string {
```